derp/derp_server_linux.go

```diff
@@ -12,40 +12,45 @@ import (
 	"tailscale.com/net/tcpinfo"
 )
 
-func (c *sclient) statsLoop(ctx context.Context) error {
+func (c *sclient) startStatsLoop(ctx context.Context) {
 	// Get the RTT initially to verify it's supported.
 	conn := c.tcpConn()
 	if conn == nil {
 		c.s.tcpRtt.Add("non-tcp", 1)
-		return nil
+		return
 	}
 	if _, err := tcpinfo.RTT(conn); err != nil {
 		c.logf("error fetching initial RTT: %v", err)
 		c.s.tcpRtt.Add("error", 1)
-		return nil
+		return
 	}
 
 	const statsInterval = 10 * time.Second
 
-	ticker, tickerChannel := c.s.clock.NewTicker(statsInterval)
-	defer ticker.Stop()
+	// Don't launch a goroutine; use a timer instead.
+	var gatherStats func()
+	gatherStats = func() {
+		// Do nothing if the context is finished.
+		if ctx.Err() != nil {
+			return
+		}
 
-statsLoop:
-	for {
-		select {
-		case <-tickerChannel:
-			rtt, err := tcpinfo.RTT(conn)
-			if err != nil {
-				continue statsLoop
-			}
+		// Reschedule ourselves when this stats gathering is finished.
+		defer func() {
+			c.s.clock.AfterFunc(statsInterval, gatherStats)
+		}()
 
-			// TODO(andrew): more metrics?
+		// Gather TCP RTT information.
+		rtt, err := tcpinfo.RTT(conn)
+		if err == nil {
 			c.s.tcpRtt.Add(durationToLabel(rtt), 1)
```

Comment:
aside: I don't get what this metric is tracking. It's not a gauge, kinda a weird counter of something that can change over time.

Reply:
I think this was essentially trying to make a histogram before we had real support for histograms; I'm going to merge for now, but we should probably make this into a "real" histogram at some point.